src/common/stack_mmap.c

```diff
@@ -0,0 +1,422 @@
+	item = tdelete(spb, &sp->root, compare_stack_size);
+	if (item == NULL)
+		D_ERROR("Size %zu not found in stack_pool %p\n", spb->stack_size, sp);
+	d_list_del(&spb->size_list);
+	sp->nb_sizes--;
+	D_FREE(spb);
+}
+
+int stack_pool_by_size_find_or_create(struct stack_pool *sp, struct stack_pool_by_size **spb,
+				      size_t size)
+{
+	void *item;
+	struct stack_pool_by_size dummy = {.stack_size = size};
+
+	item = tfind((void *)&dummy, &sp->root, compare_stack_size);
+	if (item != NULL) {
+		*spb = *((void **)item);
+		D_DEBUG(DB_MEM, "sub-pool by-size %p has been found in pool %p for size %zu\n",
+			*spb, sp, size);
+		return 0;
+	}
+
+	/* size not found, create it */
+	D_ALLOC(*spb, sizeof(struct stack_pool_by_size));
+	if (*spb == NULL)
+		return -DER_NOMEM;
+	D_INIT_LIST_HEAD(&(*spb)->stack_free_list);
+	D_INIT_LIST_HEAD(&(*spb)->size_list);
+	(*spb)->stack_size = size;
+	item = tsearch((void *)(*spb), &sp->root, compare_stack_size);
+	if (item == NULL) {
+		return -DER_NOMEM;
+	} else if (*((void **)item) == (void *)(*spb)) {
+		/* new stack size has been added to the btree */
+		sp->nb_sizes++;
+		d_list_add_tail(&(*spb)->size_list, &sp->stack_size_list);
+		D_DEBUG(DB_MEM, "sub-pool by-size %p has been created in pool %p for size %zu\n",
+			*spb, sp, size);
+	} else {
+		/* this should not happen, size finally exists! so no need new/same one */
+		D_DEBUG(DB_MEM,
+			"sub-pool by-size %p of size %zu already exists in pool %p finally..., so freeing %p\n",
+			*((void **)item), size, sp, *spb);
+		D_FREE(*spb);
+		*spb = *((void **)item);
+	}
+
+	return 0;
+}
+
+int mmap_stack_thread_create_common(struct stack_pool *sp, enum AbtThreadCreateType flag,
+				    void *arg, void (*thread_func)(void *), void *thread_arg,
+				    ABT_thread_attr attr, ABT_thread *newthread)
+{
+	ABT_thread_attr new_attr = ABT_THREAD_ATTR_NULL;
+	int rc;
+	void *stack;
+	mmap_stack_desc_t *mmap_stack_desc = NULL;
+	size_t stack_size = MMAPED_ULT_STACK_SIZE, usable_stack_size;
+	struct stack_pool_by_size *spb = NULL;
+
+	if (daos_ult_mmap_stack == false) {
+		rc = call_abt_method(arg, flag, thread_func, thread_arg,
+				     attr, newthread);
+		return rc;
+	}
+
+	if (attr != ABT_THREAD_ATTR_NULL) {
+		ABT_thread_attr_get_stack(attr, &stack, &stack_size);
+		if (stack != NULL) {
+			/* an other external stack allocation method is being
+			 * used, nothing to do
+			 */
+			rc = call_abt_method(arg, flag, thread_func, thread_arg,
+					     attr, newthread);
+			return rc;
+		}
+		if (stack_size < MMAPED_ULT_STACK_SIZE)
+			stack_size = MMAPED_ULT_STACK_SIZE;
+	} else {
+		rc = ABT_thread_attr_create(&new_attr);
+		if (rc != ABT_SUCCESS) {
+			D_ERROR("Create ABT thread attr failed: %d\n", rc);
+			return rc;
+		}
+		attr = new_attr;
+	}
+
+	/* XXX a stack is allocated from the creating XStream's stack pool
+	 * but will be freed on the running one ...
+	 */
+
+	rc = stack_pool_by_size_find_or_create(sp, &spb, stack_size);
```

Comment:
I'm wondering why we need this? It seems to me that we have one stack size (1MB) only?